File: goatcode/session.py
"""Session persistence: JSONL transcripts under ~/.goatcode/sessions/.

JSONL over SQLite keeps the dependency count at zero and memory tiny —
we only ever load the tail window we need.
"""
from __future__ import annotations

import json
import time
import uuid
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any

from .config import app_dir
from .llm import Message, ToolCall


def sessions_dir() -> Path:
    d = app_dir() / "sessions"
    d.mkdir(parents=True, exist_ok=True)
    return d


@dataclass
class Session:
    id: str
    cwd: str
    model: str
    created: float = field(default_factory=time.time)
    updated: float = field(default_factory=time.time)
    title: str = ""
    messages: list[Message] = field(default_factory=list)
    compacted_from: int = 0  # messages before this index are summarized away

    @property
    def path(self) -> Path:
        return sessions_dir() / f"{self.id}.jsonl"

    @classmethod
    def new(cls, cwd: str, model: str) -> "Session":
        return cls(id=uuid.uuid4().hex[:12], cwd=cwd, model=model)

    def save(self) -> None:
        header = {"type": "meta", "id": self.id, "cwd": self.cwd, "model": self.model,
                  "created": self.created, "updated": time.time(), "title": self.title,
                  "compacted_from": self.compacted_from}
        lines = [json.dumps(header)]
        for m in self.messages:
            entry = {"type": "message", **asdict(m)}
            lines.append(json.dumps(entry))
        self.path.write_text("\n".join(lines) + "\n", encoding="utf-8")

    @classmethod
    def load(cls, sid: str) -> "Session":
        path = sessions_dir() / f"{sid}.jsonl"
        if not path.exists():
            raise FileNotFoundError(f"no session {sid}")
        return cls._parse(path.read_text(encoding="utf-8"))

    @classmethod
    def _parse(cls, text: str) -> "Session":
        meta: dict[str, Any] = {}
        messages: list[Message] = []
        for line in text.splitlines():
            try:
                data = json.loads(line)
            except json.JSONDecodeError:
                continue
            if data.get("type") == "meta":
                meta = data
            elif data.get("type") == "message":
                calls = [ToolCall(**tc) for tc in data.get("tool_calls") or []]
                messages.append(Message(role=data.get("role", "user"),
                                        content=data.get("content") or "",
                                        tool_calls=calls,
                                        tool_call_id=data.get("tool_call_id"),
                                        name=data.get("name")))
        return cls(id=meta.get("id", "?"), cwd=meta.get("cwd", ""), model=meta.get("model", ""),
                   created=meta.get("created", 0), updated=meta.get("updated", 0),
                   title=meta.get("title", ""), messages=messages,
                   compacted_from=meta.get("compacted_from", 0))

    def context(self) -> list[Message]:
        """Messages to send to the model (skips compacted prefix)."""
        return self.messages[self.compacted_from:]

    def append(self, msg: Message) -> None:
        self.messages.append(msg)
        self.updated = time.time()


def list_sessions(limit: int = 20) -> list[dict[str, Any]]:
    rows = []
    for path in sorted(sessions_dir().glob("*.jsonl"), key=lambda p: p.stat().st_mtime, reverse=True)[:limit]:
        try:
            with path.open(encoding="utf-8") as fh:
                meta = json.loads(fh.readline())
            if meta.get("type") == "meta":
                meta["id"] = path.stem
                rows.append(meta)
        except (json.JSONDecodeError, IndexError):
            continue
    return rows


def summarize(messages: list[Message], max_chars: int = 4000) -> str:
    """Cheap deterministic compaction: last-N user/assistant text digest.

    No extra LLM call — keeps GoatCode usable on slow connections and free
    tiers. The digest is injected as a system message.
    """
    parts: list[str] = []
    total = 0
    for m in reversed(messages):
        if m.role not in ("user", "assistant") or not m.content:
            continue
        snippet = m.content.strip().replace("\n", " ")[:280]
        line = f"[{m.role}] {snippet}"
        total += len(line)
        if total > max_chars:
            break
        parts.append(line)
    return "Earlier conversation summary (oldest last):\n" + "\n".join(reversed(parts))

File: src/session.ts
/**
 * JSONL session persistence: ~/.goatcode/sessions/<id>.jsonl
 */
import { existsSync, mkdirSync, readFileSync, readdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { appDir } from "./config.ts";
import type { Message, ToolCall } from "./llm.ts";
import { textOf } from "./llm.ts";

export interface SessionData {
  id: string;
  cwd: string;
  model: string;
  title: string;
  createdAt: number;
  usage: { in: number; out: number; cacheRead?: number; cacheWrite?: number };
  /** Message-count at the start of each user turn (for /rewind). */
  turnMarks?: number[];
  messages: Message[];
  compactedFrom: number;
  /** Model-written (or digest-fallback) summary of everything before compactedFrom. */
  digest: string;
}

function sessionsDir(): string {
  const d = join(appDir(), "sessions");
  mkdirSync(d, { recursive: true });
  return d;
}

function newId(): string {
  const t = new Date();
  const pad = (n: number) => String(n).padStart(2, "0");
  // random suffix — two Session.new() calls in the same second must not collide
  const rnd = Math.random().toString(36).slice(2, 6);
  return `${t.getFullYear()}${pad(t.getMonth() + 1)}${pad(t.getDate())}-${pad(t.getHours())}${pad(t.getMinutes())}${pad(t.getSeconds())}${rnd}`;
}

export class Session {
  id: string;
  cwd: string;
  model: string;
  title = "";
  createdAt: number;
  messages: Message[] = [];
  compactedFrom = 0;
  usage: { in: number; out: number; cacheRead?: number; cacheWrite?: number } = { in: 0, out: 0 };
  digest = "";
  /** Message-count at the start of each user turn (drives /rewind). */
  turnMarks: number[] = [];

  constructor(init: Partial<SessionData> & { id: string; cwd: string; model: string }) {
    this.id = init.id;
    this.cwd = init.cwd;
    this.model = init.model;
    this.title = init.title ?? "";
    this.createdAt = init.createdAt ?? Date.now() / 1000;
    this.messages = init.messages ?? [];
    this.compactedFrom = init.compactedFrom ?? 0;
    const u = init.usage ?? { in: 0, out: 0 };
    this.usage = { in: u.in ?? 0, out: u.out ?? 0, cacheRead: u.cacheRead ?? 0, cacheWrite: u.cacheWrite ?? 0 };
    this.digest = init.digest ?? "";
    this.turnMarks = Array.isArray(init.turnMarks) ? init.turnMarks : [];
  }

  static new(cwd: string, model: string): Session {
    return new Session({ id: newId(), cwd, model });
  }

  static loadById(id: string): Session {
    const s = new Session({ id, cwd: process.cwd(), model: "" });
    s.load();
    return s;
  }

  private path(): string {
    return join(sessionsDir(), `${this.id}.jsonl`);
  }

  append(msg: Message): void {
    this.messages.push(msg);
  }

  save(): void {
    const lines = [
      JSON.stringify({
        meta: true, id: this.id, cwd: this.cwd, model: this.model,
        title: this.title, createdAt: this.createdAt,
        compactedFrom: this.compactedFrom, usage: this.usage, digest: this.digest,
        turnMarks: this.turnMarks,
      }),
      ...this.messages.map((m) => JSON.stringify(m)),
    ];
    writeFileSync(this.path(), lines.join("\n") + "\n", "utf8");
  }

  load(): void {
    if (!existsSync(this.path())) throw new Error(`no session ${this.id}`);
    const lines = readFileSync(this.path(), "utf8").split("\n").filter(Boolean);
    const msgs: Message[] = [];
    for (const line of lines) {
      const obj = JSON.parse(line);
      if (obj.meta) {
        this.cwd = obj.cwd; this.model = obj.model; this.title = obj.title ?? "";
        this.createdAt = obj.createdAt; this.compactedFrom = obj.compactedFrom ?? 0;
        this.usage = { cacheRead: 0, cacheWrite: 0, ...(obj.usage ?? {}) };
        this.digest = obj.digest ?? "";
        this.turnMarks = Array.isArray(obj.turnMarks) ? obj.turnMarks : [];
      } else {
        msgs.push(obj as Message);
      }
    }
    this.messages = msgs;
  }

  /** The window sent to the provider: everything after the compaction cut. */
  context(): Message[] {
    return this.messages.slice(this.compactedFrom);
  }
}

export function listSessions(): { id: string; model: string; title: string }[] {
  const dir = sessionsDir();
  const out: { id: string; model: string; title: string }[] = [];
  for (const f of readdirSync(dir).filter((f) => f.endsWith(".jsonl"))) {
    try {
      const first = readFileSync(join(dir, f), "utf8").split("\n", 1)[0];
      const meta = JSON.parse(first);
      if (meta.meta) out.push({ id: meta.id, model: meta.model ?? "", title: meta.title ?? "" });
    } catch { /* skip corrupt */ }
  }
  return out.sort((a, b) => b.id.localeCompare(a.id));
}

export interface SessionSummary {
  id: string;
  model: string;
  title: string;
  cwd: string;
  createdAt: number;   // unix seconds
  usage: { in: number; out: number; cacheRead?: number; cacheWrite?: number };
  /** Message-count at the start of each user turn (for /rewind). */
  turnMarks?: number[];
}

/** Every stored session's meta line — feeds the cost dashboard. */
export function allSessions(): SessionSummary[] {
  const dir = sessionsDir();
  const out: SessionSummary[] = [];
  for (const f of readdirSync(dir).filter((f) => f.endsWith(".jsonl"))) {
    try {
      const first = readFileSync(join(dir, f), "utf8").split("\n", 1)[0];
      const meta = JSON.parse(first);
      if (meta.meta) out.push({
        id: meta.id, model: meta.model ?? "", title: meta.title ?? "",
        cwd: meta.cwd ?? "", createdAt: meta.createdAt ?? 0,
        usage: { cacheRead: 0, cacheWrite: 0, ...(meta.usage ?? {}) },
        turnMarks: Array.isArray(meta.turnMarks) ? meta.turnMarks : [],
      });
    } catch { /* skip corrupt */ }
  }
  return out.sort((a, b) => b.createdAt - a.createdAt);
}

/**
 * Full-text search across saved sessions (user + assistant text). Case-
 * insensitive substring; newest-first; corrupt files skipped.
 */
export interface SessionHit {
  id: string; title: string; model: string; createdAt: number;
  snippet: string; hits: number;
}

export function searchSessions(query: string, limit = 20): SessionHit[] {
  const q = query.trim().toLowerCase();
  if (!q) return [];
  const dir = sessionsDir();
  const out: SessionHit[] = [];
  let files: string[];
  try { files = readdirSync(dir).filter((f) => f.endsWith(".jsonl")); } catch { return []; }
  for (const f of files) {
    try {
      const lines = readFileSync(join(dir, f), "utf8").split("\n").filter(Boolean);
      let title = "", model = "", createdAt = 0, id = f.replace(/\.jsonl$/, "");
      let hits = 0;
      let snippet = "";
      for (const line of lines) {
        let obj: any;
        try { obj = JSON.parse(line); } catch { continue; }
        if (obj.meta) { title = obj.title ?? ""; model = obj.model ?? ""; createdAt = obj.createdAt ?? 0; id = obj.id ?? id; continue; }
        if (obj.role !== "user" && obj.role !== "assistant") continue;
        const text = typeof obj.content === "string" ? obj.content
          : Array.isArray(obj.content) ? obj.content.filter((p: any) => p.type === "text").map((p: any) => p.text).join(" ") : "";
        const low = text.toLowerCase();
        let pos = low.indexOf(q), count = 0;
        while (pos !== -1) { count++; pos = low.indexOf(q, pos + q.length); }
        if (!count) continue;
        hits += count;
        if (!snippet) {
          const at = low.indexOf(q);
          const from = Math.max(0, at - 40);
          snippet = (from > 0 ? "…" : "") + text.slice(from, at + q.length + 60).replace(/\s+/g, " ") + "…";
        }
      }
      if (hits) out.push({ id, title, model, createdAt, snippet, hits });
    } catch { /* skip unreadable */ }
  }
  out.sort((a, b) => b.createdAt - a.createdAt);
  return out.slice(0, limit);
}

/** Cheap deterministic digest used for compaction (no extra API call). */
export function summarize(messages: Message[]): string {
  const parts: string[] = [];
  for (const m of messages.slice(0, 60)) {
    const text = textOf(m.content);
    const head = text.split("\n").slice(0, 2).join(" ").slice(0, 160);
    const img = text.includes("[image]") ? " [image]" : "";
    const tools = m.toolCalls?.map((tc: ToolCall) => `${tc.name}(${String(tc.arguments.path ?? tc.arguments.command ?? "").slice(0, 60)})`).join(", ");
    parts.push(`${m.role}: ${head}${img}${tools ? ` [called: ${tools}]` : ""}`);
  }
  return `[Summary of ${messages.length} earlier messages]\n${parts.join("\n")}`;
}
